### PR TITLE
Feature: Add strict mode

### PR DESCRIPTION
**Pull request by joehorsnell**
Note: This PR includes https://github.com/jarmo/require_all/commit/a2216e59db4493228ea0ce743430c4e24e4b0c5c from PR #20, so if you're happy with that one, I'll rebase this PR after that's merged.

Enable strict mode to immediately throw exceptions that occur during the requiring process. This makes it easy to find places where explicit requires are needed, to ensure direct dependencies are loaded. This can be done globally, by setting `RequireAll.strict_mode` to `true`, or on a per method basis by passing `strict_mode: true` to the options argument of `require_all` or `require_rel`.

By default, `require_all` will catch (and swallow) errors such as `NameError` that occur when a file is required, but keep track of the files that didn't load successfully and retry them later.

This approach, while convenient, can be dangerous and lead to subtle and/or confusing issues, as it subverts the guarantee of `require` that a given file will only be loaded once, a guarantee that code might (fairly) depend on.

For example, if an error such as a `NameError` occurs during the require of a file (because an as yet unloaded constant is referenced) that class ends up being partially defined - the constant for that class exists, but only the code up to the point of the error will have run. When that file is required again (ie when retried) the same code that executed previously will run _again_ too. This will cause any constants that were already defined in the class to be redefined (issuing warnings from Ruby) but more seriously, if another class uses that class in its partially defined state, some code that it depends on (eg methods being defined) may not exist yet, even though the constant for the class does. There is no easy way to fix this, as the only way of seeing what the result of requiring a file would be is to require it!

### CODE REVIEW

**Comment by joehorsnell** [comment]
@jarmo, I've just noticed that my PR changed a number of unrelated lines by removing trailing whitespace (apologies). I can revert that if you'd prefer - let me know.

**Comment by jarmo** [comment]
Thank you for this PR. Yeah, please remove whitespace changes before I'm going to dive into actual changes.

File: spec/fixtures/resolvable_with_strict_mode/a.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+require_relative 'c'
+
+class A < C; end
```
Comment (owst) [inline spec/fixtures/resolvable_with_strict_mode/a.rb]: These fixture files should probably all end with a trailing EOL

File: spec/require_spec.rb [hunk 2]
```diff
@@ -1,26 +1,73 @@
...
+      def with_strict_mode
+        RequireAll.strict_mode = true
+        yield
+      ensure
+        RequireAll.strict_mode = false
+      end
+
+      it "can require resolvable dependencies when there are no errors raised" do
+        with_strict_mode do
+          fixture = send(fixture_method, 'resolvable_with_strict_mode/*.rb')
```
Comment (owst) [inline spec/require_spec.rb]: Do we need this glob, given that the path will eventually be passed through to `require_all` (i.e. would `... 'resolvable_with_strict_mode/')` suffice?

File: spec/require_spec.rb [hunk 3]
```diff
@@ -1,26 +1,73 @@
...
+            send(method, fixture)
+          end.to raise_error(NameError)
+        end
+      end
+    end
+
+    context 'using options on method call' do
+      it "can require resolvable dependencies when there are no errors raised" do
+        fixture = send(fixture_method, 'resolvable_with_strict_mode/*.rb')
+        send(method, fixture, :strict_mode => true)
```
Comment (owst) [inline spec/require_spec.rb]: Can/should we use new-style symbol key syntax (`strict_mode: true`)

Comment (jarmo) [reply spec/require_spec.rb]: Ok, just decided that current [stable versions of Ruby](https://www.ruby-lang.org/en/downloads/) should be supported in which case using "new-style" is ok. Updated [CI configuration](https://github.com/jarmo/require_all/blob/master/.travis.yml) accordingly.

Still, I'd introduce that change globally in a separate PR.

Comment (owst) [reply spec/require_spec.rb]: @jarmo done! https://github.com/jarmo/require_all/pull/22

File: spec/fixtures/resolvable_with_strict_mode/a.rb (hunk 1 above)
Comment (joehorsnell) [reply spec/fixtures/resolvable_with_strict_mode/a.rb]: Yep, agree - done.

**Comment by joehorsnell** [comment]
I've done those requested follow-ups @jarmo - let me know if there's anything else.

**Comment by enkessler** [comment]
I like this idea. Strict mode would prevent the problem encountered in #8 while still allowing users to require files en masse.

**Comment by joehorsnell** [comment]
Ha - that's quite a coincidence @enkessler - I literally just added a comment re this on issue #8! 😆

**Comment by joehorsnell** [comment]
Also @jarmo / @enkessler, it would be good to get PR #20 and #22 merged, if you are happy with those? This one can then be rebased, assuming that's your policy?

**Comment by jarmo** [comment]
The more I think about it the more I like the idea of strict mode being enabled by default. As mentioned above, this would prevent many unexpected problems. Why not reveal these problems as early as possible then?

What do you guys think if there's no option **strict_mode** and there are no if statements regarding that in the code in the first place, but require_all would always raise an exception if there's problem loading files forcing developers to specify some loading order if there are such dependencies?

**Comment by joehorsnell** [comment]
Hi @jarmo - I think it's definitely sensible for strict mode to be the default, since the current approach is unsafe. However, as you point out, that's not backwards compatible,

I think it comes down to how many people are using this gem whether you do it incrementally, or just jump straight to your suggestion.

According to RubyGems, 8.2m downloads ever, and 3.8m for the current version, so not massive, but probably enough to warrant doing it incrementally, given how simple that is to do.

So, I'd suggest adding strict mode now as an option, which is a backwards compatible change (i.e. minor version bump). I can update the docs accordingly regarding the next major version on this PR?

**Comment by jarmo** [comment]
What value do you see doing it incrementally? As I see it then it is much more costly to fix bugs, which reveal themselves later at some point than to explode immediately. Also, bumping a major version should not introduce that many problems, because I hope that 99% people using this gem uses pessimistic version locking inside of their Gemfile.

I'd say that let's make code cleaner already by removing it as an option, have this functionality written into a README and into changelog and bump major version should be the next steps.

**Comment by joehorsnell** [comment]
In general it obviously just allows users to upgrade and get the benefits of other improvements, but without having to do other more invasive work (eg if they weren't encountering this issue).

But since there aren't really any other improvements, yes I think you are probably right - change it and bump the major version.

I can have a look at that if you like?

**Comment by enkessler** [comment]
I'm a fan of having `strict` as a separate (but default) mode. I like encouraging good practices (thus having it as the default) but I also like flexibility (thus leaving in the option to not care about errors). I also like download counts and I wouldn't be surprised if a significant portion of the current user base uses this gem exactly so that they don't have to worry about how their files fit. It'd be nice not to lose them.

I haven't looked at the code in a while but there should only be one place in the code where we should have to check for strict mode. That place is wherever we capture errors from loaded files currently. That being said, I wouldn't be surprised if there is more than one spot due to this being a long-lived project with many contributors. Instead of letting that be a deterrent, however, it can be viewed as an opportunity to refactor the current code so that there is just a single point where it cares if a file blows up or not. :)

I also suggest going straight to `2.0`. People who don't want to use strict can just not upgrade and people who don't have version limits in their gemfiles wouldn't be protected by an incremental upgrade anyway.

**Comment by jarmo** [comment]
That's what I was pointing out - if anyone really, for some strange reasons, like to have strict mode off, then they can use an older version. I'm pretty sure that 100% of the users who would like to have it off, have not had to deal with any strange errors in production code thus far. I, as a developer, would be happy if I have to do some extra work as to not cause any problems whatsoever in production because of a pure chance (which require_all currently is without its strict mode). However, if upgrading is a pain in the a**, then anyone can postpone it until some time in the future, but upgrading gives all developers a possibility of future-proofing their code a little.

Having any options or whatsoever just adds unnecessary complexity and I don't think there should be anyone turning strict mode off, because it is just asking for trouble at the time you least expect it.

Just have it behave as is it with any npm build tool, which concatenate JavaScript files into one for example where you have to specify order of files if there are any dependencies and then load everything else.

**Comment by jarmo** [comment]
Closing this because of #24.

**Comment by mensfeld** [comment]
Hey, just wanted to give you a headsup - we've been using require_all in Karafka for a really long time and this is a great decision (to which we will have to adapt now but still)! Keep up the great work!

**Comment by jarmo** [comment]
Thanks for these encouraging words! We all have to thank @joehorsnell for that in the end :)

**Comment by martincsimpson** [comment]
Another heads up - some deprecation notice / warning would have been nice. We've just had a production outage due to the automatic upgrading of the gem from 1.5.0 to 2.0.0 (our fault for not locking the version in Gemfile.lock) - but I thought you'd like to know that an incremental (i.e deprecation warnings) would have been better for us.

That being said, we've used this gem for years, and it's fantastic... so thanks for the hard work.

**Comment by jarmo** [comment]
Thank you for your feedback. I'm sorry to hear that you experienced problems (and a little bit surprised that you didn't experience any problems in production before of the same problem, which got fixed/changed).

It was actually pointed out by the collaborator to add a deprecation warning, but I didn't think that it's necessary because in my naivety I thought that 99% of the users already have require_all version locked with pessimistic versioning in their Gemfile and I never thought that installing into production will install latest version.

Adding a deprecation warning would have postponed releasing version 2.0.0 by an extended period of time and there would not have been still any guarantees that it would have prevented this problem to happen when let's say you or anyone else would not have deployed to production or installed latest versions between that time period. It's a tough problem to solve for everyone, really.

**Comment by mensfeld** [comment]
@jarmo No worries. It was our fault not to lock require_all in the gemspec the way it should. We're just before the new release, so your timing was great!

> and a little bit surprised that you didn't experience any problems in production before of the same problem

Well that's because Karafka is most of the time used together with Rails and it has it's own autoloading and then we don't use the require_all. Require_all in Karafka context is used in two places:

- loading internal stuff (always)
- loading user code when no other framework available (standalone)

Why we didn't have problems with it earlier? Well because we have a really good structure of code, where there are almost none dependencies in between files :)

I wrote this message to say that this move was good and the problems that the change causes are minimal comparing to the risk you might have when not loading stuff properly. If it hit production for anyone, I would strongly advise to rethink their work strategy and introduce test / staging layer. Keep up the great work!

**Comment by martincsimpson** [comment]
@jarmo Understood on your feedback, and again - you're right that we actually have pessimistic version locking but in this particular deploy the Gemfile.lock wasn't correctly recognised which caused a mass-upgrade. Unfortunately, it was just require_all that had breaking changes. 

With that being said, all is fixed now and we're back to using it as normal. Thanks for the great gem!

**Comment by jarmo** [comment]
Thanks again for your feedback, guys :) I'd have to say that I'm impressed that only require_all caused problems when its major version got upgraded. You're either upgrading your dependencies quite often or you just got damn lucky this time :D

**Pull request closed without merging**